webSockets/receiver/receiver.go

```diff
@@ -2,57 +2,85 @@ package receiver
 
 import (
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/gorilla/websocket"
 	"github.com/multiversx/mx-chain-core-go/core"
 	"github.com/multiversx/mx-chain-core-go/core/check"
 	"github.com/multiversx/mx-chain-core-go/core/closing"
 	"github.com/multiversx/mx-chain-core-go/webSockets"
-	"github.com/multiversx/mx-chain-core-go/webSockets/connection"
 	"github.com/multiversx/mx-chain-core-go/webSockets/data"
 )
 
 // ArgsReceiver holds the arguments that are needed for a receiver
 type ArgsReceiver struct {
-	Uint64ByteSliceConverter connection.Uint64ByteSliceConverter
+	Uint64ByteSliceConverter webSockets.Uint64ByteSliceConverter
 	Log                      core.Logger
 	RetryDurationInSec       int
 	BlockingAckOnError       bool
 }
 
 type receiver struct {
-	payloadHandler           webSockets.PayloadHandler
-	uint64ByteSliceConverter connection.Uint64ByteSliceConverter
-	log                      core.Logger
-	safeCloser               core.SafeCloser
-	retryDuration            time.Duration
-	blockingAckOnError       bool
+	payloadParser      webSockets.PayloadParser
+	payloadHandler     webSockets.PayloadHandler
+	log                core.Logger
+	safeCloser         core.SafeCloser
+	retryDuration      time.Duration
+	blockingAckOnError bool
+	mutex              sync.RWMutex
 }
 
 // NewReceiver will create a new instance of receiver
 func NewReceiver(args ArgsReceiver) (*receiver, error) {
-	if err := checkArgs(args); err != nil {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	payloadParser, err := webSockets.NewWebSocketPayloadParser(args.Uint64ByteSliceConverter)
+	if err != nil {
 		return nil, err
 	}
 
 	return &receiver{
-		log:                      args.Log,
-		uint64ByteSliceConverter: args.Uint64ByteSliceConverter,
-		retryDuration:            time.Duration(args.RetryDurationInSec) * time.Second,
-		blockingAckOnError:       args.BlockingAckOnError,
-		safeCloser:               closing.NewSafeChanCloser(),
-		payloadHandler:           webSockets.NewNilPayloadHandler(),
+		log:                args.Log,
+		retryDuration:      time.Duration(args.RetryDurationInSec) * time.Second,
+		blockingAckOnError: args.BlockingAckOnError,
+		safeCloser:         closing.NewSafeChanCloser(),
+		payloadHandler:     webSockets.NewNilPayloadHandler(),
+		payloadParser:      payloadParser,
 	}, nil
 }
 
+func checkArgs(args ArgsReceiver) error {
+	if check.IfNil(args.Log) {
+		return core.ErrNilLogger
+	}
+	if check.IfNil(args.Uint64ByteSliceConverter) {
+		return data.ErrNilUint64ByteSliceConverter
+	}
+	if args.RetryDurationInSec == 0 {
+		return data.ErrZeroValueRetryDuration
+	}
+	return nil
+}
+
 // SetPayloadHandler will set the payload handler
-func (r *receiver) SetPayloadHandler(handler webSockets.PayloadHandler) {
+func (r *receiver) SetPayloadHandler(handler webSockets.PayloadHandler) error {
+	if check.IfNilReflect(handler) {
```

Comment:
please use check.IfNil. handler should extend the NilChecker interface

Reply:
done